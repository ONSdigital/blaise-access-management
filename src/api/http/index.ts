import axios from "axios";

export * from "./users";
export * from "./roles";

export function handleAxiosError(error: unknown): { status: number; message: string, error: unknown } {
    if (axios.isAxiosError(error)) {
        return {
            status: error.response?.status || 500,
            message: error.response?.data.message || "Axios error occurred with no specific message",
            error: error
        };
    } else {
        return {
            status: 500,
            message: "An unknown error occurred, please raise a ServiceNow call if this continues: " + error,
            error: error
        };
    }
}