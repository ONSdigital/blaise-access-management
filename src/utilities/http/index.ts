export * from "./users";
export * from "./roles";
export * from "./auth";
