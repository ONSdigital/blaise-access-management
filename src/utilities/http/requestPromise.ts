import {AuthManager} from "blaise-login-react-client";

type PromiseResponse = [number, any];

function requestPromiseJson(method: string, url: string, body: any = null, headers: any = {}): Promise<PromiseResponse> {
    const authManager = new AuthManager();

    return new Promise((resolve: (object: PromiseResponse) => void, reject: (error: string) => void) => {
        fetch(url, {
            "method": method,
            "body": body,
            "headers": Object.assign({}, headers, authManager.authHeader())
        })
            .then(async response => {
                response.json().then(
                    data => (resolve([response.status, data]))
                ).catch((error) => {
                    resolve([response.status, "Read JSON failed"]);
                });
            })
            .catch(err => {
                console.error(err);
                reject(err);
            });
    });
}

type PromiseResponseList = [boolean, []];

function requestPromiseJsonList(method: string, url: string, body: any = null): Promise<PromiseResponseList> {
    const authManager = new AuthManager();
    return new Promise((resolve: (object: PromiseResponseList) => void, reject: (error: string) => void) => {
        fetch(url, {
            "method": method,
            "body": body,
            "headers": authManager.authHeader()
        })
            .then(async response => {
                response.json().then(
                    data => {
                        if (response.status === 200) {
                            if (!Array.isArray(data)) {
                                resolve([false, []]);
                            }
                            resolve([true, data]);
                        } else if (response.status === 404) {
                            resolve([true, data]);
                        } else {
                            resolve([false, []]);
                        }
                    }
                ).catch((error) => {
                    console.error(error);
                    resolve([false, []]);
                });
            })
            .catch(err => {
                reject(err);
            });
    });
}

export {requestPromiseJson, requestPromiseJsonList};
