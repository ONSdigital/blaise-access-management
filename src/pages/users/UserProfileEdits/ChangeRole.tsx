import React from "react";
import { ONSButton, ONSErrorPanel, ONSLoadingPanel, ONSPanel } from "blaise-design-system-react-components";
import { ChangeEvent, ReactElement, useEffect, useState } from "react";
import { Navigate, useLocation, useParams } from "react-router-dom";
import Breadcrumbs from "../../../Components/Breadcrumbs";
import { BreadcrumbItem } from "../../../Interfaces";
import { RedirectWithData, UserRouteParams } from "../../../Interfaces/usersPage";
import { getAllRoles, patchUserRolesAndPermissions } from "../../../api/http";
import { UserRole } from "blaise-api-node-client";
import Form from "../../../Components/form";
import UserSignInErrorPanel from "../../../Components/UserSignInErrorPanel";

export default function ChangeRole(): ReactElement {
    const { user: viewedUsername }: UserRouteParams = useParams() as unknown as UserRouteParams;
    const { state } = useLocation();
    const { currentUser, viewedUserDetails } = state || { currentUser: null, viewedUserDetails: null };
    const [role, setRole] = useState<string>(viewedUserDetails?.data?.role ?? "");
    const [roleList, setRoleList] = useState<UserRole[]>([]);
    const [redirectWithData, setRedirectWithData] = useState<RedirectWithData>({ redirect: false, visible: false, message: "", statusType: "" });
    const [setError, setSetError] = useState<string | null>(null);
    const [setLoading, setSetLoading] = useState<boolean>(true);

    const getRoleList = async () => {
        try {
            const [_success, roleList] = await getAllRoles();
            setRoleList(roleList);
            setSetLoading(false);
        } catch (error) {
            setSetError("Failed to fetch roles list, please try again");
            setSetLoading(false);
        }
    };

    const handleChangeRole = (e: ChangeEvent<HTMLSelectElement>) => {
        setRole(e.target.value);
        return;
    };

    const changeBlaiseUserRolesAndServerParks = async () => {
        if (!viewedUserDetails || !viewedUserDetails.data) {
            console.log("Viewed user details is undefined or null");
            return;
        }

        if (role === viewedUserDetails.role) {
            console.log("User already has role: ", role);
            return;
        }

        if (roleList.some((userRole) => userRole.name === role)) {
            const res = await patchUserRolesAndPermissions(viewedUsername, role);
            setRedirectWithData({ redirect: true, visible: true, message: res?.message || "", statusType: res?.status === 500 ? "error" : "success" });
        } else {
            window.alert(`Invalid role: ${role}`);
            console.log("Invalid Role:", role);
            return;
        }
    };

    const breadcrumbList: BreadcrumbItem[] = [
        { link: "/", title: "Home" },
        { link: "/users/", title: "Manage users" },
        { link: `/users/${viewedUsername}`, title: "View user", state: { currentUser } }
    ];

    useEffect(() => {
        getRoleList();
    }, []);

    if (!currentUser || !viewedUserDetails || setError) {
        return setError ? (<ONSPanel status="error">{setError}</ONSPanel>) : (<UserSignInErrorPanel/>);
    }

    return (
        <>
            {
                redirectWithData.redirect && <Navigate
                    to={{ pathname: `/users/${viewedUsername}` }}
                    state={{
                        currentUser,
                        updatedPanel: {
                            visible: redirectWithData.visible,
                            message: redirectWithData.message,
                            status: redirectWithData.statusType
                        }
                    }}
                />
            }
            <Breadcrumbs BreadcrumbList={breadcrumbList} />
            {
                setLoading ? (
                    <ONSLoadingPanel />
                ) : (
                    <main id="main-content" className="ons-page__main ons-u-mt-no">
                        <h1 className="ons-u-mb-l">Change current role for user <em>{viewedUsername}</em></h1>
                        <h2 className="ons-u-mb-l">Current role: <em>{viewedUserDetails?.data?.role ?? "N/A"}</em></h2>
                        <Form onSubmit={() => changeBlaiseUserRolesAndServerParks()}>
                            <p className="ons-field">
                                <label className="ons-label" htmlFor="new-user-role">New Role
                                </label>
                                <select value={role} id="new-user-role" name="select-role" className="ons-input ons-input--select "
                                    onChange={(e) => handleChangeRole(e)}>
                                    {
                                        roleList.map((option: UserRole) => {
                                            return (<option key={option.name} value={option.name}>{option.name}</option>);
                                        })
                                    }
                                </select>
                            </p>
                            <ONSButton
                                label={"Save"}
                                primary={true}
                                submit={true} />
                        </Form>
                    </main>
                )
            }
        </>
    );
}
