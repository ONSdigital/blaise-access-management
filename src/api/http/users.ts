import axios from "axios";
import { requestPromiseJson, requestPromiseJsonList } from "./requestPromise";
import { User, NewUser } from "blaise-api-node-client";
import { AuthManager } from "blaise-login-react/blaise-login-react-client";
import { GetUserResponse, GetUsersListResponse, PatchUserRoleResponse } from "../../Interfaces/usersPage";
import { handleAxiosError } from ".";

async function getAllUsers(): Promise<GetUsersListResponse> {
    try {
        const url = "/api/users";
        const [success, data] = await requestPromiseJsonList<User>("GET", url);
        return [success, data] as GetUsersListResponse;
    } catch (error) {
        return [false, []] as GetUsersListResponse;
    }
}

// NOTE: Catch block is different due to the use of AsyncContent and AsyncRequest
async function getUser(user: string): Promise<GetUserResponse> {
    try {
        const authManager = new AuthManager();
        const url = `/api/users/${user}`;
        const headers = authManager.authHeader();

        const res = await axios.get(url, { headers });

        return {
            status: res.status,
            message: res.data.message,
            data: res.data.data
        };

    } catch (error) {
        const errorObj = handleAxiosError(error);
        console.error(errorObj.message, { ...errorObj, data: {}, error: error });
        throw Error(errorObj.message);
    }
}

function addNewUser(newUser: NewUser): Promise<boolean> {
    const url = "/api/users";

    return new Promise((resolve: (object: boolean) => void) => {

        if (!newUser.password) {
            resolve(false);
            return;
        }

        const formData = new FormData();
        formData.append("name", newUser.name);
        formData.append("password", newUser.password);
        formData.append("role", newUser.role);

        requestPromiseJson("POST", url, formData).then(([status]) => {
            if (status === 200 || status === 201) {
                resolve(true);
            } else {
                resolve(false);
            }
        }).catch(() => {
            resolve(false);
        });
    });
}

function deleteUser(username: string): Promise<boolean> {
    const url = "/api/users";

    const headers = {
        "user": username
    };

    return new Promise((resolve: (object: boolean) => void) => {

        requestPromiseJson("DELETE", url, null, headers).then(([status]) => {
            if (status === 204) {
                resolve(true);
            } else {
                resolve(false);
            }
        }).catch(() => {
            resolve(false);
        });
    });
}

// BLAIS5-3816
async function patchUserRolesAndPermissions(user: string, role: string, currentServerParks: string[], currentDefaultServerPark: string): Promise<PatchUserRoleResponse> {
    try {
        const authManager = new AuthManager();
        const url = `/api/users/${user}/rolesAndPermissions`;
        const headers = authManager.authHeader();
        const data = { role, currentServerParks, currentDefaultServerPark };

        const res = await axios.patch(url, data, { headers });

        return {
            status: res.status,
            message: res.data.message
        };
    } catch (error) {
        const errorObj = handleAxiosError(error);
        console.error(errorObj.message, errorObj);
        return errorObj;
    }
}

export { getAllUsers, getUser, addNewUser, deleteUser, patchUserRolesAndPermissions };
