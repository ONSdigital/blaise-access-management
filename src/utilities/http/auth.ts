import {requestPromiseJson} from "./requestPromise";
import {User} from "../../../Interfaces";

type authLoginUserResponse = [boolean, User | null];

function loginUser(username: string, password: string): Promise<authLoginUserResponse> {
    console.log(`Call to login user with user ${username}`);
    const url = `/api/login/${username}`;

    const headers = {
        "password": password,
    };

    return new Promise((resolve: (object: authLoginUserResponse) => void) => {
        requestPromiseJson("GET", url, null, headers).then(([status, data]) => {
            console.log(`Response from login user: Status ${status}, data ${data}`);
            if (status === 200) {
                resolve([true, data]);
            } else if (status === 403) {
                resolve([true, null]);
            } else {
                resolve([false, null]);
            }
        }).catch((error: Error) => {
            console.error(`Response from login user Failed: Error ${error}`);
            resolve([false, null]);
        });
    });
}

export {loginUser};
