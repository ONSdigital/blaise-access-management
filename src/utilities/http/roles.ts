import {requestPromiseJsonList} from "./requestPromise";
import {Role} from "../../../Interfaces";

type getRolesListResponse = [boolean, Role[]];

function getAllRoles(): Promise<getRolesListResponse> {
    console.log("Call to getAllRoles");
    const url = "/api/roles";

    return new Promise((resolve: (object: getRolesListResponse) => void) => {
        requestPromiseJsonList("GET", url).then(([success, data]) => {
            console.log(`Response from get all roles ${(success ? "successful" : "failed")}, data list length ${data.length}`);
            resolve([success, data]);
        }).catch((error: Error) => {
            console.error(`Response from get all roles Failed: Error ${error}`);
            resolve([false, []]);
        });
    });
}

export {getAllRoles};
