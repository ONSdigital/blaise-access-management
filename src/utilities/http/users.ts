import {requestPromiseJson, requestPromiseJsonList} from "./requestPromise";
import {User} from "../../../Interfaces";

type getUsersListResponse = [boolean, User[]];

function getAllUsers(): Promise<getUsersListResponse> {
    console.log("Call to getAllUsers");
    const url = "/api/users";

    return new Promise((resolve: (object: getUsersListResponse) => void) => {
        requestPromiseJsonList("GET", url).then(([success, data]) => {
            console.log(`Response from get all users ${(success ? "successful" : "failed")}, data list length ${data.length}`);
            resolve([success, data]);
        }).catch((error: Error) => {
            console.error(`Response from get all users Failed: Error ${error}`);
            resolve([false, []]);
        });
    });
}

function addNewUser(newUser: User): Promise<boolean> {
    console.log("Call to addNewUser");

    const url = "/api/users";

    return new Promise((resolve: (object: boolean) => void) => {

        if (newUser.password === undefined) {
            resolve(false);
            return;
        }

        const formData = new FormData();
        formData.append("name", newUser.name);
        formData.append("password", newUser.password);
        formData.append("role", newUser.role);

        requestPromiseJson("POST", url, formData).then(([status, data]) => {
            console.log(`Response from add new user: Status ${status}, data ${data}`);
            if (status === 201) {
                resolve(true);
            } else {
                resolve(false);
            }
        }).catch((error: Error) => {
            console.error(`Response from add new user Failed: Error ${error}`);
            resolve(false);
        });
    });
}

export {getAllUsers, addNewUser};
