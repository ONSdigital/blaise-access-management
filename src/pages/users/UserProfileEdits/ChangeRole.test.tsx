/**
 * @jest-environment jsdom
 */

import React from "react";
import { render, cleanup, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import ChangeRole from "./ChangeRole";
import { MemoryRouter, useParams } from "react-router-dom";
import { getAllRoles, patchUserRolesAndPermissions } from "../../../api/http";
import userEvent from "@testing-library/user-event";
import { UserRole } from "blaise-api-node-client";

jest.mock("react-router-dom", () => ({
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    ...jest.requireActual("react-router-dom"),
    useParams: jest.fn()
}));

jest.mock("../../../api/http", () => ({
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    ...jest.requireActual("../../../api/http"),
    getAllRoles: jest.fn(),
    patchUserRolesAndPermissions: jest.fn()
}));

const mockRoles = [
    { name: "DST", description: "DST User" },
    { name: "BDSS", description: "BDSS User" },
    { name: "IPS Manager", description: "IPS Field Interviewer User" },
    { name: "IPS Field Interviewer", description: "IPS Manager User" },
    { name: "IPS Researcher", description: "IPS Researcher User" },
    { name: "Editor", description: "Editor User" },
    { name: "Editor Manager", description: "Editor Manager User" },
    { name: "TO Appointments", description: "TO Appointments User" },
    { name: "TO Manager", description: "TO Manager User" },
    { name: "TO Interviewer", description: "TO Interviewer User" },
    { name: "SEL", description: "SEL User" },
    { name: "Welsh Speaker", description: "Welsh Speaker User" }
];

const mockUserDetails = {
    data: { role: "DST" },
    name: "testUser",
    role: "DST",
    serverParks: ["gusty"],
    defaultServerPark: "gusty"
};

const mockState = {
    pathname: `users/${mockUserDetails.name}/change-role`,
    state: { currentUser: "currentUser", viewedUserDetails: mockUserDetails }
};

beforeEach(() => {
    (getAllRoles as jest.Mock).mockResolvedValue([true, mockRoles]);
    (useParams as jest.Mock).mockReturnValue({ user: mockUserDetails.name });
});

afterEach(() => cleanup());

describe("ChangeRole Component (with state management)", () => {
    it("matches the snapshot", async () => {
        (patchUserRolesAndPermissions as unknown as jest.Mock).mockResolvedValue({ message: "Role updated successfully", status: 200 });

        const { asFragment } = render(
            <MemoryRouter initialEntries={[mockState]}>
                <ChangeRole />
            </MemoryRouter>
        );

        // Wait for state update
        await act(async () => {});

        expect(asFragment()).toMatchSnapshot();
    });

    it("renders and displays the correct initial role", async () => {
        const { findByText, findAllByText } = render(
            <MemoryRouter initialEntries={[mockState]}>
                <ChangeRole />
            </MemoryRouter>
        );

        // Wait for state update
        await act(async () => {});

        expect(await findByText(/Current role:/i)).toBeVisible();
        expect((await findAllByText(/DST/i))[0]).toBeVisible();
    });

    it("updates role upon form submission", async () => {
        global.confirm = jest.fn(() => true);

        const { findByText, findByRole } = render(
            <MemoryRouter initialEntries={[mockState]}>
                <ChangeRole />
            </MemoryRouter>
        );

        // Wait for state update
        await act(async () => {});

        const select = await findByRole("combobox");
        const saveButton = await findByText("Save");

        act(() => {
            userEvent.selectOptions(select, ["IPS Field Interviewer"]);
        });

        expect(await findByText(/IPS Field Interviewer/i)).toBeVisible();

        act(() => {
            userEvent.click(saveButton);
        });

        // Wait for state update
        await act(async () => {});

        expect(patchUserRolesAndPermissions).toHaveBeenCalledTimes(1);
        expect(patchUserRolesAndPermissions).toHaveBeenCalledWith("testUser", "IPS Field Interviewer");
    });

    it("displays an error message when fetching roles fails", async () => {
        (getAllRoles as jest.Mock).mockRejectedValue(new Error("Failed to fetch roles"));

        const { findByText } = render(
            <MemoryRouter initialEntries={[mockState]}>
                <ChangeRole />
            </MemoryRouter>
        );

        // Wait for state update
        await act(async () => {});

        expect(await findByText(/Failed to fetch roles list, please try again/i)).toBeVisible();
    });
});
