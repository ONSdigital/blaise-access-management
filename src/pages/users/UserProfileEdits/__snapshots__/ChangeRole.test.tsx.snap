// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeRole Component matches the snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="Breadcrumb"
    class="ons-breadcrumb"
  >
    <ol
      class="ons-breadcrumb__items ons-u-fs-s"
    >
      <li
        class="ons-breadcrumb__item"
        id="breadcrumb-0"
      >
        <a
          class="ons-breadcrumb__link"
          href="/"
        >
          Home
        </a>
        <svg
          class="ons-svg-icon"
          focusable="false"
          viewBox="0 0 8 13"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
            transform="translate(-5.02 -1.59)"
          />
        </svg>
      </li>
      <li
        class="ons-breadcrumb__item"
        id="breadcrumb-1"
      >
        <a
          class="ons-breadcrumb__link"
          href="/users/"
        >
          Manage users
        </a>
        <svg
          class="ons-svg-icon"
          focusable="false"
          viewBox="0 0 8 13"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
            transform="translate(-5.02 -1.59)"
          />
        </svg>
      </li>
      <li
        class="ons-breadcrumb__item"
        id="breadcrumb-2"
      >
        <a
          class="ons-breadcrumb__link"
          href="/users/undefined"
        >
          View user
        </a>
        <svg
          class="ons-svg-icon"
          focusable="false"
          viewBox="0 0 8 13"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.74,14.28l-.57-.56a.5.5,0,0,1,0-.71h0l5-5-5-5a.5.5,0,0,1,0-.71h0l.57-.56a.5.5,0,0,1,.71,0h0l5.93,5.93a.5.5,0,0,1,0,.7L6.45,14.28a.5.5,0,0,1-.71,0Z"
            transform="translate(-5.02 -1.59)"
          />
        </svg>
      </li>
    </ol>
  </nav>
  <main
    class="ons-page__main ons-u-mt-no"
    id="main-content"
  >
    <h1
      class="ons-u-mb-l"
    >
      Change current role for user 
      <em />
    </h1>
    <h2
      class="ons-u-mb-l"
    >
      Current role: 
      <em>
        DST
      </em>
    </h2>
    <form
      class="ons-u-mt-m"
    >
      <p
        class="ons-field"
      >
        <label
          class="ons-label"
          for="new-user-role"
        >
          New Role
        </label>
        <select
          class="ons-input ons-input--select "
          id="new-user-role"
          name="select-role"
        />
      </p>
      <button
        class="ons-btn "
        data-testid="button"
        type="submit"
      >
        <span
          class="ons-btn__inner"
        >
          Save
        </span>
      </button>
    </form>
  </main>
</DocumentFragment>
`;
